clustered/server/src/main/java/org/ehcache/clustered/server/internal/messages/EhcacheSyncMessageCodec.java

```diff
@@ -257,20 +258,22 @@ private EhcacheSyncMessage decodeDataSync(ByteBuffer message) {
   }
 
   private Map<Long, Chain> decodeChainMapEntries(StructDecoder<Void> decoder) {
-    Map<Long, Chain> chainMap = new HashMap<>();
-
     StructArrayDecoder<? extends StructDecoder<?>> entriesDecoder = decoder.structs(CHAIN_MAP_ENTRIES_SUB_STRUCT);
+
     if (entriesDecoder != null) {
-      for (int i = 0; i < entriesDecoder.length(); i++) {
+      int len = entriesDecoder.length();
+      Map<Long, Chain> chainMap = (len < 16) ? new HashMap<>() : new HashMap<>((int) ((float) len / 0.75f + 1.0f));
```

Comment:
Why default when less than 16? Also, no need to apply a factor to the initial capacity. This is done by the HashMap. Just put the length. We are not expecting new entries to be added don't we?

Reply:
Sadly the HashMap pre-size doesn't take the loadfactor in to account:
```java
Map<Integer, Integer> map = new HashMap<>(16);

Field tableField = HashMap.class.getDeclaredField("table");
tableField.setAccessible(true);

for (int i = 0; i < 16; i++) {
  map.put(i, i);
  System.out.println("map.size() = " + map.size() + " : map.table.length = " + ((Object[]) tableField.get(map)).length);
}
```
gives:
```
map.size() = 1 : map.table.length = 16
map.size() = 2 : map.table.length = 16
map.size() = 3 : map.table.length = 16
map.size() = 4 : map.table.length = 16
map.size() = 5 : map.table.length = 16
map.size() = 6 : map.table.length = 16
map.size() = 7 : map.table.length = 16
map.size() = 8 : map.table.length = 16
map.size() = 9 : map.table.length = 16
map.size() = 10 : map.table.length = 16
map.size() = 11 : map.table.length = 16
map.size() = 12 : map.table.length = 16
map.size() = 13 : map.table.length = 32
map.size() = 14 : map.table.length = 32
map.size() = 15 : map.table.length = 32
map.size() = 16 : map.table.length = 32
```

Reply:
Huh? It kinda does. The size goes to 32 after 12. Which is about 75% of 16. And then, they use a power of 2 for everything.

Reply:
Okay, but thats my point.  Ramesh is trying to size the HashMap such that he can fit 'N' things without having to rehash the table at any point.  To achieve that he can't just do `new HashMap<>(N)` since that constructor just sizes the initial table, and the map obeys a default load factor of 0.75.  Instead he needs (at a minimum) to size it at `ceiling(N/0.75)` which he has approximated as `(N/0.75)+1`.